### PR TITLE
Hydra train skeleton

### PR DESCRIPTION
**Pull request by KayLamerigts**
Possible to override all kinds of shit from the command line

```bash
PYTHONPATH=. python moths/scripts/train.py data.fake_data=true +trainer.instance.limit_train_batches=10 +trainer.instance.limit_val_batches=10 trainer.instance.max_epochs=10
```

### CODE REVIEW

File: moths/data_module.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+from moths.config import resolve_config_path
+from moths.datasets import (ConcatImageFolderDataset, FakeData,
+                            LabelMapImageFolder)
+
+log = logging.getLogger(__name__)
+
+
+@dataclass
+class DataConfig(DictConfig):
+    data_paths: List[str]
```
Comment (lmdehaas) [inline moths/data_module.py:22]: From downstream use it looks like this can also be List[Path]

File: moths/data_module.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+    train_transforms: List[Any]
```
Comment (lmdehaas) [inline moths/data_module.py:24]: Isn't there a Transforms baseclass from pytorch we can use here?

File: moths/model.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+from omegaconf import DictConfig
+from torch import nn
+from torchvision import models
+
+log = logging.getLogger(__name__)
+
+
+@dataclass
+class ModelConfig(DictConfig):
+    zoo_name: str = "resnet50"
```
Comment (lmdehaas) [inline moths/model.py:13]: Why define defaults here if we use something like Hydra? I can imagine this default could lead to silent errors later on.

File: moths/config/trainer/dev.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    id: null # pass correct id to resume experiment!
+    log_model: False
+    prefix: ""
+    job_type: "train"
+    group: ""
+    tags: [ ]
+
+
+callbacks:
+  - _target_: pytorch_lightning.callbacks.ModelCheckpoint
```
Comment (lmdehaas) [inline moths/config/trainer/dev.yaml:40]: Is this still needed when we also log model checkpoints to W&B?

File: moths/data_module.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                    str(data_path),
+                    label_map=self.label_map,
+                    is_valid_file=lambda p: abspath(p) in self.valid_paths,
+                    transform=transform,
+                )
+                datasets.append(ds)
+
+        return ConcatImageFolderDataset(datasets=datasets)
+
+    def setup(self, stage: Optional[str] = None):
```
Comment (lmdehaas) [inline moths/data_module.py:96]: `stage` isn't used anywhere, probably an artifact?

File: moths/data_module.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            log.info(
+                f"Using {[str(resolve_config_path(p)) for p in self.config.data_paths]}."
+            )
+
+    def prepare_data(self):
+        valid_data_path = resolve_config_path(self.config.valid_path_file)
+        log.info(f"Using valid data {str(valid_data_path)} ...")
+
+        with valid_data_path.open("r") as f:
+            self.valid_paths = set([p.strip() for p in f.readlines()])
```
Comment (lmdehaas) [inline moths/data_module.py]: nit: 
```suggestion
            self.valid_paths = {p.strip() for p in f.readlines()}
```

File: moths/datasets.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        else:
+            sample_idx = idx - self.cumulative_sizes[dataset_idx - 1]
+        return self.datasets[dataset_idx][sample_idx]
+
+    def __len__(self):
+        return sum(len(d) for d in self.datasets)
+
+    @property
+    def targets(self):
+        return sum([d.targets for d in self.datasets], [])
```
Comment (lmdehaas) [inline moths/datasets.py:91]: What exactly do you sum here, and is this something we might want to unittest to be sure?

File: moths/data_module.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        self.val_dataset = Subset(full_test_dataset, val_indices)
+        self.test_dataset = Subset(full_test_dataset, test_indices)
+
+    def train_dataloader(self):
+        return DataLoader(
+            self.train_dataset,
+            batch_size=self.config.batch_size,
+            num_workers=self.config.num_workers,
+            pin_memory=self.config.pin_memory,
+            shuffle=True,
```
Comment (lmdehaas) [inline moths/data_module.py:133]: Is this also seeded?

File: moths/datasets.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+
+class ConcatImageFolderDataset(Dataset):
+    def __init__(self, datasets: List[Union[ImageFolder, FakeData]]):
+        """Stricter implementation of ConcatDataset for ImageFolders such that it can use targets."""
+        super(ConcatImageFolderDataset, self).__init__()
+        self.datasets = list(datasets)  # make super sure it is a list for order
+        self.cumulative_sizes = ConcatDataset.cumsum(self.datasets)
+
+    def __getitem__(self, idx: int) -> Tuple[Any, ...]:
```
Comment (lmdehaas) [inline moths/datasets.py:78]: Can you add a docstring here? To me it's not instantly clear what this `__getitem__` does. I guess it gets the item from the corresponding dataset, depending on the index?

File: moths/logging.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+            hp[f"data/{name}_transform_{i}"] = f"{cfg['_target_'].split('.')[-1]}"
+
+    hp["model"] = config.model
+    hp["model/num_parameters"] = sum(p.numel() for p in lit_module.parameters())
+
+    # TODO: add everything to here, probably flattened?
+    # TODO: how to deal with all these class instances?
+
+    # it says it cant, but it works fine with a dict
+    trainer.logger.log_hyperparams(hp)  # type: ignore
```
Comment (lmdehaas) [inline moths/logging.py:41]: Doesn't this log the dict as a str or something, instead of separate key-value pairs?

File: moths/logging.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+    # hacky trick found on the internet
```
Comment (lmdehaas) [inline moths/logging.py:43]: What does this trick do exactly?

File: moths/scripts/valid_data_paths.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
+from pathlib import Path
```
Comment (lmdehaas) [inline moths/scripts/valid_data_paths.py:1]: How and when is this script used?

**Review by lmdehaas — COMMENTED** [review]
Very nice! Some points:
- Might be too soon to start nagging about this stuff but maybe we should update the readme with training usage?
- Also maybe too soon to start nagging, but should we add unittest (e.g. for the `targets` property)
- I don't see a new run using this code in W&B. Did you already try running it?

File: moths/config/trainer/dev.yaml (hunk 4 above)
Comment (KayLamerigts) [reply moths/config/trainer/dev.yaml:40]: No, not needed. Maybe nice to have.

I think just defensive of me since I don't know W&B :sweat_smile:

File: moths/data_module.py (hunk 1 above)
Comment (KayLamerigts) [reply moths/data_module.py:22]: Hydra doesn't accept Path unfortunately

File: moths/data_module.py (hunk 2 above)
Comment (KayLamerigts) [reply moths/data_module.py:24]: Weirdly no, transforms don't have a parent.

File: moths/data_module.py (hunk 6 above)
Comment (KayLamerigts) [reply moths/data_module.py]: wow, what is this suggested magic

File: moths/data_module.py (hunk 5 above)
Comment (KayLamerigts) [reply moths/data_module.py:96]: yeah, function signature from parent.

It calls once with "train", "validation", "test", and finally None, if I am not mistaken.

Does that mean it is executed four times now? Maybe yes :smile:

File: moths/data_module.py (hunk 8 above)
Comment (KayLamerigts) [reply moths/data_module.py:133]: Yes, pytorch lignthing has `seed_everything` which should seed numpy, pytorch, and lightning all for everything.

File: moths/datasets.py (hunk 9 above)
Comment (KayLamerigts) [reply moths/datasets.py:78]: Correct, will add.

File: moths/datasets.py (hunk 7 above)
Comment (KayLamerigts) [reply moths/datasets.py:91]: targets is a list, sum with starting element `[]` concats the lists.

```python
>>> sum([[0, 1], [2, 3]], [])
[0, 1, 2, 3]
```
The reason for building this class is to get an unified targets from the concatenated datasets that can be used to stratify the concatenation.

File: moths/logging.py (hunk 10 above)
Comment (KayLamerigts) [reply moths/logging.py:41]: Here in the bottom under `Config` you can see the result

https://wandb.ai/butterflies/dev/runs/h2ouux75/overview

File: moths/logging.py (hunk 11 above)
Comment (KayLamerigts) [reply moths/logging.py:43]: Makes sure the hyperparams are logged only once, tbh, I saw someone else do it and just did it without knowing if for us it is also logged twice :innocent:

File: moths/model.py (hunk 3 above)
Comment (KayLamerigts) [reply moths/model.py:13]: from when I started, will remove

File: moths/scripts/valid_data_paths.py (hunk 12 above)
Comment (KayLamerigts) [reply moths/scripts/valid_data_paths.py:1]: You remember that starting a train run took 99 hours because it has to loop over all files and check if they are valid for every run? I thought it would be nice if we do that only once. 

Run once, created file, give file in config. Then override the `is_valid_file` function to check if it is in this list.

File: moths/config/trainer/dev.yaml (hunk 4 above)
Comment (lmdehaas) [reply moths/config/trainer/dev.yaml:40]: Right, we can also keep it for now, both fine to me!

File: moths/data_module.py (hunk 6 above)
Comment (lmdehaas) [reply moths/data_module.py]: Python ❤️

File: moths/scripts/valid_data_paths.py (hunk 12 above)
Comment (lmdehaas) [reply moths/scripts/valid_data_paths.py:1]: Awesomeeeee

File: moths/datasets.py (hunk 7 above)
Comment (lmdehaas) [reply moths/datasets.py:91]: Cool, didn't know that!

**Comment by KayLamerigts** [comment]
@lmdehaas it was not nice of me to make such a big boy

I will close the PR and split it up in smaller parts, that is nicer I think. I will refer back here for your comments.
